File: week4/binarySearch.js
process.stdin.setEncoding('utf8');

var data = ""

function binarySearch(sortedArray, secondArrayInteger) {
  var low = 0;
  var high = sortedArray.length - 1;

  while (low <= high) {
    mid = (Math.floor((high - low) / 2)) + low;
    element = sortedArray[mid];
    if (element < secondArrayInteger) {
      low = mid + 1;
    } else if (element > secondArrayInteger) {
      high = mid - 1;
    } else {
      return mid;
    }
  }
  return -1;

  // var mid = Math.floor((high - low)/ 2) + low;
  // var element = sortedArray[mid];

  // if (low > high) return -1;

  // if (element < secondArrayInteger) {
  //   return binarySearch(sortedArray, secondArrayInteger, mid + 1, high)
  // } else if (element > secondArrayInteger) {
  //   return binarySearch(sortedArray, secondArrayInteger, low, mid - 1)
  // } else {
  //   return mid
  // }
};

process.stdin.on('end', function() {
  var lines = data.split("\n")

  var firstArray = [];
  var secondArray = [];
  var answerArray = [];

  var firstArray = lines[0]
    .split(' ')
    .map( function(elem){
      return parseInt(elem,10);
    })
    .slice(1);

  var secondArray = lines[1]
    .split(' ')
    .map( function(elem){
      return parseInt(elem,10);
    })
    .slice(1);

  // var low = 0;
  // var high = firstArray.length - 1;

  for (var i = 0; i < secondArray.length; i++) {
    answerArray.push(binarySearch(firstArray, secondArray[i]));
  }

  var output = answerArray.join(" ")

  console.log(output)
  process.exit()
})

process.stdin.on('readable', function(){
  new_data = process.stdin.read();
  if (new_data !== null) {
    data = data + new_data
  }
})

File: week4/binarySearch.cs
using System;

namespace BinarySearch
{
  class Program
  {
    static void Main()
    {
      string[] input;

      input = Console.ReadLine().Split(' ');
      var n = Convert.ToInt32(input[0]);
      var SortedArray = new int[n];
      for(var i = 0; i < n; i++)
        SortedArray[i] = Convert.ToInt32(input[i + 1]);

      input = Console.ReadLine().Split(' ');
      var m = Convert.ToInt32(input[0]);
      var SearchTerms = new int[m];
      for(var i = 0; i < m; i++)
        SearchTerms[i] = Convert.ToInt32(input[i + 1]);

      var results = new int[m];
      for(var i = 0; i < m; i++)
        results[i] = BinarySearchWrapper(SortedArray, SearchTerms[i]);

      Console.WriteLine(String.Join(" ", results));
    }

    private static int BinarySearchWrapper(int[] SortedArray, int SearchTerm)
    {
      return BinarySearch(SortedArray, SearchTerm, 0, SortedArray.Length - 1);
    }
    private static int BinarySearch(int[] SortedArray, int SearchTerm, int low, int high)
    {
      if(high < low)
        return -1;

      var mid = low + ((high - low) / 2);

      if(SearchTerm == SortedArray[mid])
        return mid;

      if(SearchTerm < SortedArray[mid])
        return BinarySearch(SortedArray, SearchTerm, low, mid - 1);

      if(SearchTerm > SortedArray[mid])
        return BinarySearch(SortedArray, SearchTerm, mid + 1, high);

      return -1;
    }
  }
}
